service/controller/v5/resource/instance/deployment.go

```diff
@@ -35,34 +62,64 @@ func (r Resource) newDeployment(ctx context.Context, obj providerv1alpha1.AzureC
 		workerNodes = append(workerNodes, n)
 	}
 
-	cc, err := controllercontext.FromContext(ctx)
+	containerName := key.BlobContainerName()
+	groupName := key.ResourceGroupName(obj)
+	storageAccountName := key.StorageAccountName(obj)
+	masterBlobName := key.BlobName(obj, key.PrefixMaster())
+	workerBlobName := key.BlobName(obj, key.PrefixWorker())
+
+	storageAccountsClient, err := r.getStorageAccountsClient(ctx)
 	if err != nil {
 		return azureresource.Deployment{}, microerror.Mask(err)
 	}
-	err = cc.Validate()
+
+	keys, err := storageAccountsClient.ListKeys(ctx, groupName, storageAccountName)
 	if err != nil {
 		return azureresource.Deployment{}, microerror.Mask(err)
 	}
 
-	masterCloudConfig, err := cc.CloudConfig.NewMasterCloudConfig(obj)
+	if len(*(keys.Keys)) == 0 {
+		return azureresource.Deployment{}, microerror.Maskf(executionFailedError, "storage account key's list is empty")
+	}
+	primaryKey := *(((*keys.Keys)[0]).Value)
+
+	masterBlobURL, err := blobclient.GetBlobURL(ctx, masterBlobName, containerName, storageAccountName, primaryKey, cc.ContainerURL)
```

Comment:
Is it secure to expose this URL in deployment template? Won't it make it available in the whole subscription? Or only this instance is authorized to access the data with this URL?

Reply:
We are not exposing it. We render small cloudconfig and put it into secure data field of the instance. Small clousconfig has this secure url for main blob (https://github.com/giantswarm/azure-operator/pull/457/files#diff-a07238412aec949b2b2c30e032093adfR92) , which appended to small ignition during boot